Article for 7.13.0 (#2450)
babel/website: website/blog/2021-02-22-7.13.0.md

layout: post
title:  "7.13.0 Released: Records and Tuples, granular compiler assumptions, and top-level targets"
author: Nicolò Ribaudo
authorURL: https://twitter.com/NicoloRibaudo
date:   2021-01-01 0:00:00
categories: announcements
share_text: "Babel 7.13.0 Released"
---

We just published Babel 7.13.0, the first minor release of 2021!

This release includes some important features to `@babel/core`: a `targets` option (similar to `@babel/preset-env`'s one) that can be shared across different plugins, an `assumptions` option to precisely tune your configuration to produce a smaller compilation output, and support for plugins and presets written using native ECMAScript modules.

Babel now supports transforming the [Records and Tuples](https://github.com/tc39/proposal-record-tuple) ECMAScript proposal, which brings immutable data structures and structural equality to JavaScript, and parsing the [Module Blocks](https://github.com/tc39/proposal-js-module-blocks) proposal.

Additionally, we added support for some new Flow and TypeScript features.

You can read the whole changelog [on GitHub](https://github.com/babel/babel/releases/tag/v7.12.0).

<!-- truncate -->

### Funding Update

> We have joined the "GitHub Sponsors for organizations" program, so you can now sponsor us on GitHub 😊.

Reviewer: ```suggestion
We have joined the "GitHub Sponsors for organizations" program, so you can now sponsor us directly through GitHub 😊.
```

<iframe src="https://github.com/sponsors/babel/button" title="Sponsor Babel on GitHub" height="35" width="116" style="border: 0; margin: auto; display: block"></iframe>

> These funds are used to pay one full-time and three part-time maintainers, to sustain the project developement, and ensure long-term improvements both regarding stability and new features.

Reviewer: ```suggestion
These funds are used to support our team's (currently one full-time and three part-time maintainers) continued efforts in improving stability and developing new features.
```

> All of our donations and expenses are transparent on our [Open Collective](https://opencollective.com/babel) page, where we will also track donations coming from GitHub Sponsors.

Reviewer: ```suggestion
Both our donations and expenses are openly tracked through our [Open Collective](https://opencollective.com/babel) page, where we will also track donations coming from GitHub Sponsors.
```

> We welcome donations both from individual and from companies: if your company is interested in becoming a Gold tier sponsor ($1k/month) and wants to discuss more please reach out to [team@babeljs.io](mailto:team@babeljs.io)!

Reviewer: ```suggestion
We welcome donations from both individual and companies: if your company is interested in becoming a Gold tier sponsor ($1k/month) and wants to discuss more, please reach out to [team@babeljs.io](mailto:team@babeljs.io)!
```

## Highlights

### Top-level `targets` option ([#12189](https://github.com/babel/babel/pull/12189), [RFC](https://github.com/babel/rfcs/pull/2))

> `targets` is the option that made `@babel/preset-env` the most powerful preset we have ever had: it can automatically choose what syntax to transform and what polyfills to inject based on your target browsers.

Reviewer: ```suggestion
The `targets` option is the true power behind `@babel/preset-env`: it allows users the ability to specify their target environments which allows the preset to automatically choose which syntax to transform and what polyfills to inject.
```

> However, other plugins can also benefit from knowing your targets. For example, you currently have to specify the targets twice if you are using our [new polyfill plugins](https://github.com/babel/babel-polyfills). With the new options you can lift it up to the top-level, specifying it only once:

Reviewer: ```suggestion
Since releasing `@babel/preset-env`, we've learned that plugins _themselves_ can also benefit from knowing your targets. Currently, this can a bit cumbersome, since you'd have to specify your targets twice (for example, if you are using our [new polyfill plugins](https://github.com/babel/babel-polyfills). By introducing `targets` as a top level option, you now only have to specify your targets once:
```

> `@babel/preset-env`'s `targets` option allows users to specify their target environments, automatically choosing which syntax to transform and what polyfills to inject.
> 
> Since releasing `@babel/preset-env`, we've learned that plugins _themselves_ can also benefit from knowing your targets. Currently, this can a bit cumbersome, as you have to specify your targets twice (for example, if you are using our [new polyfill plugins](https://github.com/babel/babel-polyfills)). By introducing `targets` as a top level option, you now only have to specify your targets once:

Author: ```suggestion
`@babel/preset-env`'s `targets` option allows users to specify their target environments, automatically choosing which syntax to transform and what polyfills to inject. Since releasing `@babel/preset-env`, we've learned that plugins _themselves_ can also benefit from knowing your targets. Currently, this can a bit cumbersome, as you have to specify your targets twice (for example, if you are using our [new polyfill plugins](https://github.com/babel/babel-polyfills)). By introducing `targets` as a top level option, you now only have to specify your targets once:
```

> Since releasing `@babel/preset-env`, we've learned that plugins _themselves_ can also benefit from knowing your targets. Currently, this can a bit cumbersome, since you'd have to specify your targets twice (for example, if you are using our [new polyfill plugins](https://github.com/babel/babel-polyfills). By introducing `targets` as a top level option, you now only have to specify your targets once:

Reviewer 2: ```suggestion
Since releasing `@babel/preset-env`, we've learned that plugins _themselves_ can also benefit from knowing your targets. Currently, this can a bit cumbersome, as you have to specify your targets twice (for example, if you are using our [new polyfill plugins](https://github.com/babel/babel-polyfills)). By introducing `targets` as a top level option, you now only have to specify your targets once:
```

<table>
<tr>
<th>Old configuration</th>
<th>New configuration</th>
</tr>
<tr>
<td>

```json
{
  "presets": [
    ["@babel/preset-env", {
      "targets": ">1%, not ie 11"
    }]
  ],
  "plugins": [
    ["polyfill-es-shims", {
      "targets": ">1%, not ie 11"
    }]
  ]
}
```

</td>
<td>

```json
{

>   "targets": ">1%, not ie 11"

Reviewer 3: ```suggestion
  "targets": ">1%, not ie 11",
```

  "presets": ["@babel/preset-env"],
  "plugins": ["polyfill-es-shims"]
}

```

</td>
</tr>
</table>

> We recommended converting configs to use the top-level `targets` option since:

Reviewer: ```suggestion
We recommended converting your Babel config to use the new top-level `targets` option since:
```

> 1. It better handles the [`esmodules: true`](TODO: Link) target. (it's intersected with the other targets, rather than replacing them)

Reviewer: ```suggestion
1. It has better handling of the [`esmodules: true`](TODO: Link) target. (it's intersected with the other targets, rather than replacing them)
```

> 2. We are exploring how to use it in more plugins to generate more optimized output

Reviewer: ```suggestion
2. As we introduce additional `targets`-related support in our plugins, you'll automatically benefit from more optimized output!
```

> 1. It has better handling of the [`esmodules: true`](TODO: Link) target. (it's intersected with the other targets, rather than replacing them)

Author: ```suggestion
1. It has better handling of the [`esmodules: true`](https://babel.dev/docs/en/options#targetsesmodules) target. (it's intersected with the other targets, rather than replacing them)
```

This will link to https://deploy-preview-2416--babel.netlify.app/docs/en/options#targetsesmodules

> 1. It has better handling of the [`esmodules: true`](https://babel.dev/docs/en/options#targetsesmodules) target. (it's intersected with the other targets, rather than replacing them)

Author: Yup, it will be correct after merging https://github.com/babel/website/pull/2416

> You can read the details about this new option in its [RFC](TODO: Link).

Author: ```suggestion
You can read the details about this new option in its [RFC](https://github.com/babel/rfcs/blob/master/rfcs/0002-top-level-targets.md).
```

> > 🔮 In the future we might explore moving `@babel/preset-env` into `@babel/core` so that you don't have to install one additional package to get started with Babel. This new option can be seen as the first step in that direction!

Reviewer: ```suggestion
> 🔮 In the future we might explore moving `@babel/preset-env` into `@babel/core` so that you don't have to install an additional package to get started with Babel. This new option can be seen as the first step in that direction!
```

### Top-level `assumptions` option ([#12219](https://github.com/babel/babel/pull/12219), [RFC](https://github.com/babel/rfcs/pull/5))

> Many of our plugins have a `loose` option, which enables the compiler to make some assumptions about the code you are writing. It uses these to ignore certain edge cases and generate a smaller or faster output.

Reviewer: ```suggestion
Many of our plugins have a `loose` option, which enables Babel to generate smaller/faster output by making certain assumptions about your code and ignoring certain edge cases.
```

> However, `loose` has some problems: the term itself is not really descriptive of how it affects the compiler behaivior and sometimes must be set the same across different plugins to make sure that everything is compiled correctly.

Reviewer: ```suggestion
However, `loose` has some problems that lead to user frustration/confusion: the term itself does nothing to describe how it affects the compiler's behavior, and worse, sometimes it must be set the same across different plugins to ensure everything compiles.
```

> For this reason, we are introducing a new set of `assumptions` that you can set independently from the plugins you are using: a plugin can be influcenced by multiple assumptions and an assumption can influence multiple plugins.

Reviewer: ```suggestion
To help fix these problems, we've added a new top level option: `assumptions`! Similar to the new `targets` option, every plugin will now receive the assumptions you've enabled, eliminating the need to set the option individually. This is really valuable since a plugin can be influenced by multiple assumptions, and an assumption can influence multiple plugins.
```

> > :warning: This is an advanced functionality. Carefully choose which assumptions you enable, because they are not spec-compliant and might break your code if violated.

Reviewer: ```suggestion
> :warning: This is **advanced** functionality. Please be careful when enabling assumptions, because they are _not_ spec-compliant and may break your code in unexpected ways.
```

> To help fix these problems, we've added a new top level option: `assumptions`! Similar to the new `targets` option, every plugin will now receive the assumptions you've enabled, eliminating the need to set the option individually. This is really valuable since a plugin can be influenced by multiple assumptions, and an assumption can influence multiple plugins.

Reviewer 2: I think the problem is described well, but I don't think it's clear what an "assumption" is at this point. The examples below help clear this up a bit, but I think it might be useful to provide a little bit more of an explanation in the intro. i.e. Why are they called assumptions? What are we making assumptions about?

Author: ```suggestion
To help fix these problems, we've added a new top level option to tell Babel which assumptions it can make about your code: `assumptions`! Similar to the new `targets` option, every plugin will now receive the assumptions you've enabled, eliminating the need to set the option individually. This is really valuable since a plugin can be influenced by multiple assumptions, and an assumption can influence multiple plugins.
```

Is this better? I don't really know how to explain what an assumption is :sweat_smile:

Reviewer 2: Yeah, that's a tough concept...I do think this is an improvement!

For example, when transforming optional chaining and nullish coalescing, Babel will generate this output by default:

<table>
<tr>
<th>Input</th>
<th>Output</th>
</tr>
<tr>
<td>

```js
let user, points;

let score = points ?? 0;
let name = user?.name;
```

</td>
<td>

```js
let user, points;

let score = points !== null && points !== void 0 ? points : 0;
let name = user === null || user === void 0 ? void 0 : user.name;
```

</td>
</tr>
</table>

> However, you can enable the `noDocumentAll` assumption to tell Babel "I'm never using the `document.all`, so you can compile without worrying about it":

Reviewer: ```suggestion
However, enabling the `noDocumentAll` assumption tells Babel "I'm never using `document.all`, so you can compile without worrying about it":
```

```json
{
  "targets": "firefox 70",
  "assumptions": { "noDocumentAll": true },
  "presets": ["@babel/preset-env"]
}
```

<table>
<tr>
<th>Input</th>
<th>Output</th>
</tr>
<tr>
<td>

```js
let user, points;

let score = points ?? 0;
let name = user?.name;
```

</td>
<td>

```js
let user, points;

let score = points != null ? points : 0;
let name = user == null ? void 0 : user.name;
```

</td>
</tr>
</table>

> You can read the full list of assumptions [in the documentation](https://babeljs.io/assumptions), and you can individually enable or disable them to see how they affect the compiled output.

Reviewer: ```suggestion
Check out the full list of assumptions [over in our documentation](https://babeljs.io/assumptions), where you can individually enable or disable them to see how they affect the compiled output.
```

### Records and Tuples support ([#12145](https://github.com/babel/babel/pull/12145))

> Thanks to a collaboration with [Bloomberg](https://www.techatbloomberg.com/), Babel now supports transforming the "Records and Tuples" stage 2 [proposal](https://github.com/tc39/proposal-record-tuple).

Reviewer 2: ```suggestion
Thanks to a collaboration with [Bloomberg](https://www.techatbloomberg.com/), Babel now supports transforming the ["Records and Tuples" stage 2 proposal](https://github.com/tc39/proposal-record-tuple).
```

> The Babel plugin transforms the records and tuples syntax using the global `Record` and `Tuple` functions:

Reviewer 2: ```suggestion
The Babel plugin transforms records and tuples syntax using the global `Record` and `Tuple` functions:
```

<table>
<tr>
<th>Input</th>
<th>Output</th>
</tr>
<tr>
<td>

```js
let data = #{
  name: "Babel",
  ids: #[1, 2, 3]
};
```

</td>
<td>

```js
let data = Record({
  name: "Babel",
  ids: Tuple(1, 2, 3)
});
```

</td>
</tr>
</table>

This means that you will need to load a polyfill for those global functions, such as [`@bloomberg/record-tuple-polyfill`](https://www.npmjs.com/package/@bloomberg/record-tuple-polyfill).

> In order to enable this transformer, you need to add `@babel/plugin-proposal-record-and-tuple` to your configuration.

Reviewer: ```suggestion
In order to enable this transform, you need to add `@babel/plugin-proposal-record-and-tuple` to your configuration.
```

### New Flow features ([#12193](https://github.com/babel/babel/pull/12193), [#12234](https://github.com/babel/babel/pull/12234))

> Babel 7.12.0 two new Flow features:

Reviewer: ```suggestion
Babel 7.12.0 has support for two new Flow features:
```

> This means that you will need to load a polyfill for those global functions, such as [`@bloomberg/record-tuple-polyfill`](https://www.npmjs.com/package/@bloomberg/record-tuple-polyfill).

Reviewer 2: Some additional context/links to documentation around using 3rd party polyfills might be useful here. As a reader, some questions I might have are:

1. What does "load a polyfill" mean?
1. How do I use Babel with a 3rd party polyfill?
1. Why isn't this covered by `targets`?

Author: ```suggestion
This means that you will need to load a polyfill for those global functions, such as [`@bloomberg/record-tuple-polyfill`](https://www.npmjs.com/package/@bloomberg/record-tuple-polyfill), either by importing it in your code or by adding it as an entrypoint to your bundler's configuration.

> **NOTE**: No engine currently supports records and tuples, so you always need to load the polyfill.
```

Does this make it clearer?

Reviewer 2: Yes!

- `this` type annotation in functions, that allow you to specify the type of the `this` object as if it was a parameter

  ```js
  function getPerson(this: Database, id: string): Person {
    this instanceof Database; // true
  }
  ```

- `enum` declarations with unknown members

  ```js
  enum PetKind {
    Dog,
    Cat,
    Snake,
    ...
  }
  ```

### TypeScript 4.2 support ([#12628](https://github.com/babel/babel/pull/12628))

[TypeScript 4.2](https://devblogs.microsoft.com/typescript/announcing-typescript-4-2-rc/) is currently RC, and it supports some new syntax features such as abstract constructor signatures.

You can read more about the changes in [the TypeScript release post](https://devblogs.microsoft.com/typescript/announcing-typescript-4-2-beta/).

### Automatic ES modules in `@babel/runtime` ([#12632](https://github.com/babel/babel/pull/12632))

> `@babel/runtime` contains all the Babel runtime helpers in two different formats: CommonJS modules and ECMAScript modules.

Reviewer: ```suggestion
`@babel/runtime` contains all of the Babel runtime helpers in both CommonJS and ECMAScript module formats.
```

> [TypeScript 4.2](https://devblogs.microsoft.com/typescript/announcing-typescript-4-2-rc/) is currently RC, and it supports some new syntax features such as abstract constructor signatures.

Reviewer 2: Should we clarify what "RC" means?

Author: We can just delete it, tomorrow it will be released as stable anyway.
```suggestion
[TypeScript 4.2](https://devblogs.microsoft.com/typescript/announcing-typescript-4-2-rc/) supports some new syntax features such as abstract constructor signatures.
```

### TypeScript 4.2 support ([#12628](https://github.com/babel/babel/pull/12628))

> [TypeScript 4.2](https://devblogs.microsoft.com/typescript/announcing-typescript-4-2-rc/) is currently RC, and it supports some new syntax features such as abstract constructor signatures.

Author: ```suggestion
[TypeScript 4.2](https://devblogs.microsoft.com/typescript/announcing-typescript-4-2-rc/) is currently RC (it will become stable on the 23rd, but we beated them by one day!) , and it supports some new syntax features such as abstract constructor signatures.
```

wdyt, assuming that we actually release on Monday 22nd? :joy:
